doc/cpp_domdec.rst

```diff
@@ -0,0 +1,186 @@
+
+Decomposition
+-------------
+
+Documentation for the data structures used to describe domain decompositions.
+
+.. cpp:namespace:: arb
+
+.. cpp:enum-class:: backend_kind
+
+    Used to indicate which hardware backend to use for running a :cpp:class:`cell_group`.
+
+    .. cpp:enumerator:: multicore
+
+        Use multicore backend.
+
+    .. cpp:enumerator:: gpu
+
+        Use gpu back end.
+
+        .. Note::
+            Can only be set if the :cpp:class:`cell_group` type supports the GPU backend.
+
+.. cpp:class:: domain_decomposition
+
+    Describes a domain decomposition. Solely responsible for describing the
+    distribution of cells across cell groups and domains.
+    Holds cell group descriptions (:cpp:member:`groups`) for cells assigned to
+    the local domain, and a helper function (:cpp:member:`gid_domain`) used to
+    look up which domain a cell has been assigned to.
+    Also has meta-data about the number of cells in the global model, and the
+    number of domains over which the model is destributed.
+
+    A load balancing algorithm generates the domain decomposition, for example see
+    :cpp:func:`partition_load_balance`.
+
+    .. cpp:function:: bool is_local_gid(cell_gid_type gid) const
+
+        Tests whether the cell with :cpp:var:`gid` is on the local domain.
+
+    .. cpp:member:: std::function<int(cell_gid_type)> gid_domain
+
+        A callback function that returns the domain id that a cell assigned to
+        (using global identifier :cpp:var:`gid`).
+        Set by the load balancing algorithm that generates the domain decomposition.
+
+    .. cpp:member:: int num_domains
+
+        Number of domains that the model is distributed over.
+
+    .. cpp:member:: int domain_id
+
+        The index of the local domain.
+        Always 0 for non-distributed models, and corresponds to the MPI rank
+        for distributed runs.
+
+    .. cpp:member:: cell_size_type num_local_cells
+
+        Total number of cells in the local domain.
+
+    .. cpp:member:: cell_size_type num_global_cells
+
+        Total number of cells in the global model
+        (sum of :cpp:member:`num_local_cells` over all domains).
+
+    .. cpp:member:: std::vector<group_description> groups
+
+        Descriptions of the cell groups on the local domain.
+        See :cpp:class:`group_description`.
+
+.. cpp:class:: group_description
+
+    The indexes of a set of cells of the same kind that are group together in a
+    cell group in a :cpp:class:`arb::simulation`.
+
+    .. cpp:function:: group_description(cell_kind k, std::vector<cell_gid_type> g, backend_kind b)
+
+        Constructor.
+
+    .. cpp:member:: const cell_kind kind
+
+        The kind of cell in the group.
+
+    .. cpp:member:: const std::vector<cell_gid_type> gids
+
+        The gids of the cells in the cell group, **sorted in ascending order**.
+
+    .. cpp:member:: const backend_kind backend
+
+        The back end on which the cell group is to run.
```

Comment:
When discussing consistency, it should be made explicit that the group descriptions are non-overlapping and their gids exhaust the set of local gids.

Reply:
See above.